lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,454 @@
+      forallRefersTo(visitingVal, [&](Node visitingNode) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "\n Erasing visiting : " << getName(visitingNode));
+        visitingSet.erase(visitingNode);
+        return success();
+      });
+      // if (auto op = visitingVal.getDefiningOp())
+      //   op->emitRemark(" is part of a combinational cycle");
     }
+  }
 
-    LLVM_DEBUG({
-      for (const auto &i1 : portPaths)
-        for (const auto &i2 : i1.second)
-          for (const auto &i3 : i2.second)
-            llvm::dbgs() << "\n node :" << i1.first << "," << i2.first
-                         << " is connected to " << i3.val << "," << i3.fieldId;
-    });
-    return success();
+  void handleConnects(Value dst, SmallVector<Node> &inputArgFields) {
+    auto pathsToOutPort = [&](Node dstNode) {
+      if (!dstNode.getValue().isa<BlockArgument>())
+        return success();
+      for (auto inArg : inputArgFields) {
+        portPaths[inArg].insert(dstNode);
+      }
+      return success();
+    };
+    forallRefersTo(dst, pathsToOutPort);
   }
 
-  void printError(Value node, ValueOpPair &childNode) {
-    if (!printCycle || currentPath.back() == node) {
-      printCycle = false;
+  void setValRefsTo(Value val, Node ref) {
+    if (!val || !ref)
       return;
-    }
-    if (childNode.second)
-      TypeSwitch<Operation *>(childNode.second)
-          .Case<InstanceOp>([&](InstanceOp ins) {
-            size_t portNum = 0, childPortNum = 0;
-            // Get the port index for `node.val`.
-            for (const auto &p : llvm::enumerate(ins->getOpResults())) {
-              if (p.value() == childNode.first)
-                childPortNum = p.index();
-              if (p.value() == node)
-                portNum = p.index();
-            }
-            auto remark = ins.emitRemark("instance is part of a combinational "
-                                         "cycle, instance port number '")
-                          << childPortNum << "' has a path from port number '"
-                          << portNum << "'";
-            FieldRef fOut = getFieldRef(node);
-            FieldRef fIn = getFieldRef(childNode.first);
-            auto inName = getFieldName(fIn).first;
-            auto outName = getFieldName(fOut).first;
-            remark << ", " << inName << " <- " << outName;
-          })
-          .Case<MemOp>([&](MemOp mem) {
-            mem.emitRemark("memory is part of a combinational cycle");
-          })
-          .Case<SubfieldOp, SubindexOp, SubaccessOp>([&](auto op) {})
-          .Default([&](auto owner) {
-            owner->emitRemark(
-                "this operation is part of the combinational cycle");
-          });
+    fieldToVals[ref].insert(val);
+    valRefersTo[val].insert(ref);
+    if (auto arg = ref.getValue().dyn_cast<BlockArgument>())
+      blockArgToFields[arg].insert(ref.getFieldID());
+  }
+
+  void forallRefersTo(Value val, std::function<LogicalResult(Node refNode)> f,
```

Comment:
clang-tidy is complaining here, could/should this be `llvm::function_ref` instead?

Reply:
It was complaining for a const, but `function_ref` sounds better updated it.